### PR TITLE
Json2csv update

### PR DESCRIPTION
**Pull request by DumbMachine**
Aims to solve #1334 and replaces the stale or #1410

### CODE REVIEW

**Comment by henrykironde** [comment]
@DumbMachine this looks good. Can we add some tests for` json2csv`. We can create some sample dataset in `test_integration`. These should have edge cases that we talked about.

File: retriever/lib/engine_tools.py [hunk 1]
```diff
@@ -101,37 +103,74 @@ def reset_retriever(scope="all", ask_permission=True):
...
     subprocess.call(['rm', '-r', input_file])
     return output_file
 
 
+def walker(dictionary, row_key=None, header_values=None, rows=[], normalize=False):
+    """
+    Both name and columns cant be none
+    """
+    #  Handles the simple case, where row_key and column_key are not required
+    if not row_key and not header_values:
```
Comment (apoorvaeternity) [inline retriever/lib/engine_tools.py:139]: DeMorgan's Law  - not (A or B)

File: retriever/lib/engine_tools.py [hunk 2]
```diff
@@ -101,37 +103,74 @@ def reset_retriever(scope="all", ask_permission=True):
...
+        if isinstance(dictionary, dict):
+            # rows = pd.Series(dictionary)
+            rows = pd.DataFrame([dictionary]).values
+            return rows
+        elif isinstance(dictionary, list):
+            rows = pd.DataFrame(dictionary, columns=header_values).values
+            return rows
+
+    if isinstance(dictionary, dict):
+        if header_values and (header_values in list(dictionary.keys()) or header_values == list(dictionary.keys())):
```
Comment (apoorvaeternity) [inline retriever/lib/engine_tools.py:149]: Use `header_values in dictionary` instead of `header_values in list(dictionary.keys())`. Hash lookup is faster.

Comment (DumbMachine) [reply retriever/lib/engine_tools.py:149]: header_values is a `list` that is why `header_values in dictionary` will not work.

Comment (DumbMachine) [reply retriever/lib/engine_tools.py:154]: header_values is a `list` that is why `header_values in dictionary` will not work.

**Comment by DumbMachine** [comment]
Hmm that's weird, my travis ci passed for my [branch](https://travis-ci.com/github/DumbMachine/retriever/builds/159395685) but not for this pr.

File: retriever/lib/engine_tools.py (hunk 2 above)
Comment (apoorvaeternity) [reply retriever/lib/engine_tools.py:154]: I may be wrong but if it's a list looking at the code it looks like you are checking if a list exists as one of the dictionary keys. I think even `header_values in list(dictionary.keys())`  will not work then. Can you add a comment showing the values of `header_values` and `dictionary.keys()` here.

Comment (DumbMachine) [reply retriever/lib/engine_tools.py]: `['id', 'firstname', 'surname', 'motivation', 'share'] dict_keys(['id', 'firstname', 'surname', 'motivation', 'share'])`
It seems that you are right, the logic I have used here is wrong. I have updated the code to use `sets` instead.

File: retriever/lib/engine_tools.py [hunk 3]
```diff
@@ -151,7 +151,9 @@ def walker(dictionary, row_key=None, header_values=None, rows=[], normalize=Fals
             return rows
 
     if isinstance(dictionary, dict):
-        if header_values and (header_values in list(dictionary.keys()) or header_values == list(dictionary.keys())):
+        print(header_values, dictionary.keys())
+        print()
+        if header_values and (set(dictionary.keys()).issubset(header_values) or header_values == list(dictionary.keys())):
```
Comment (apoorvaeternity) [inline retriever/lib/engine_tools.py:156]: I guess you want to do `set(header_values).issubset(dictionary.keys())`. Also, after this `header_values == list(dictionary.keys())` may not be required.

File: retriever/lib/engine_tools.py [hunk 4]
```diff
@@ -101,37 +103,77 @@ def reset_retriever(scope="all", ask_permission=True):
...
+    if not (row_key or header_values):
+        if isinstance(dictionary, dict):
+            rows = pd.DataFrame([dictionary]).values
+            return rows
+        elif isinstance(dictionary, list):
+            rows = pd.DataFrame(dictionary, columns=header_values).values
+            return rows
+
+    if isinstance(dictionary, dict):
+        if header_values and (set(dictionary.keys()).issubset(header_values):
```
Comment (apoorvaeternity) [inline retriever/lib/engine_tools.py:151]: `set(header_values).issubset(dictionary.keys())`

File: retriever/lib/engine_tools.py [hunk 5]
```diff
@@ -101,37 +103,77 @@ def reset_retriever(scope="all", ask_permission=True):
...
+            if normalize:
+                rows.extend(json_normalize(dict(i for i in dictionary.items() if i[0] in header_values)).values)
+            else:
+                rows.extend([dict(i for i in dictionary.items() if i[0] in header_values)])
+
+        elif row_key and row_key in list(dictionary.keys()):
```
Comment (apoorvaeternity) [inline retriever/lib/engine_tools.py:157]: `dictionary.get(row_key)`

File: test/test_retriever.py [hunk 6]
```diff
@@ -26,6 +27,12 @@
     **{"tables": test_engine.table, "name": "test"})
 test_engine.opts = {'database_name': '{db}_abc'}
 
+json2csv_datasets = [
+    # test_name, json_data, header_values, row_key, expected
+    ("simple_json", """[ {"User": "Alex", "Country": "US", "Age": "25"} ]""", ['User','Country','Age'], None, ['User,Country,Age', 'Alex,US,25']),
+    ("nested_json", """{"prizes":[{"year":"2019","category":"chemistry","laureates":[{"id":"976","firstname":"John","surname":"Goodenough","motivation":"text shorted","share":"3"}]}]}""", ["id", "firstname", "surname", "motivation", "share"], 'prizes', ['id,firstname,surname,motivation,share', '976,John,Goodenough,text shorted,3']),
+]
```
Comment (henrykironde) [inline test/test_retriever.py:34]: Could we add some edge cases like this data
```
[
{"User": "Alex", "id": "US1", "Age": "25", "kt": "2.0", "qt": "1.00"},
{"User": "Tom", "id": "US2", "Age": "20", "kt": "0.0", "qt": "1.0"},
{"User": "Dan", "id": "44", "Age": "2", "kt": "0", "qt": "1"},
{"User": "Kim", "id": "654", "Age": "", "kt": "", "qt": ""}
]
```

Comment (DumbMachine) [reply test/test_retriever.py:35]: This has been added, please check.

**Comment by DumbMachine** [comment]
image
Wierd, `yapf` doesn't show any warnings for me.

**Comment by henrykironde** [comment]
Install the branch using `pip install . -U` and run yapf.
Yapf will help you format your code. make sure you are up to date with master. 
Then run yapf with -i to format the new changes. 
The -d option will display the errors.

**Comment by DumbMachine** [comment]
I don't seem to be getting the same errors as the build
image

**Comment by henrykironde** [comment]
Ref: #1507
